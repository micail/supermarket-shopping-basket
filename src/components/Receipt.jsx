import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Totals from './Totals';
import { Button } from 'react-bootstrap';
import { deleteItem } from '../redux/actions/receiptActions';

const Receipt = () => {
  const priceList = useSelector((state) => state.priceList);
  const receipt = useSelector((state) => state.receipt);

  const dispatch = useDispatch();
  const removeItem = (position) => {
    dispatch(deleteItem(position));
  };

  const standardItem = (item, index) => {
    const itemName = priceList[item][0];
    const itemPrice = priceList[item][1];
    return (
      <li key={index} className="list-group-item border-0">
        {itemName}
        {' '}
        {itemPrice}
        {' '}
        <Button className="m-1" variant="warning" size="xs" key={itemName} type="button" onClick={() => removeItem(index)}>
          Delete
        </Button>
      </li>
    );
  };

  const weightItem = (item, index) => {
    const priceListKey = item[0];
    console.log(priceList)
    const itemName = priceList[priceListKey][0];
    const price = priceList[priceListKey][1];
    const unit = priceList[priceListKey][2];

    const itemCost = item[2];
    const itemWeight = item[1];
    return (
      <li key={index}>
        {itemName}
        <br />
        {itemWeight}
        {' '}
        {price}
        /
        {unit}
        {' '}
        {itemCost}
      </li>
    );
  };

  const CheckoutSummary = () => {
    return (<ul className="col-xs-12 list-group">
      {
        receipt.map((item, index) => {
          return typeof item === 'number'
            ? standardItem(item, index)
            : weightItem(item, index);
        })
      }
    </ul>);
  }

  return (
    <div className="Receipt col-md-6">
      <CheckoutSummary />
      <Totals />
    </div>
  );
};

export default Receipt;
