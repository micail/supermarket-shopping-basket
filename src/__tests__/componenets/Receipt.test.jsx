import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import Receipt from '../../components/Receipt';

const mockStore = configureMockStore();

describe('<Receipt />', () => {
  const store = mockStore({
    priceList: [
      ['Beans', 0.50],
      ['Coke', 0.50],
      ['Oranges', 1.99, 'kg'],
    ],
    receipt: [0, 0, 0, 1, 1, [2, 0.3, 0.6]],
    totals: {
      subTotal: 3.5,
      savings: [['Beans 3 for 2', -0.5], ['Coke 2 for Â£1', -0.4]],
      totalSavings: -0.9,
      totalToPay: '2.60',
    },
  });

  const wrapper = mount(<Provider store={store}><Receipt /></Provider>);

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render list of items based on priceList occurrence in the receipt', () => {
    expect(wrapper.find('.Receipt').children().length).toEqual(2);
  });
});
