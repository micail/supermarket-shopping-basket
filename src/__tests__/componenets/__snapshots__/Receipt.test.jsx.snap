// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Receipt /> should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Receipt>
    <ul
      className="Receipt"
    >
      <li
        key="0"
      >
        Beans
         
        £0.5
      </li>
      <li
        key="1"
      >
        Beans
         
        £0.5
      </li>
      <li
        key="2"
      >
        Beans
         
        £0.5
      </li>
      <li
        key="3"
      >
        Coke
         
        £0.5
      </li>
      <li
        key="4"
      >
        Coke
         
        £0.5
      </li>
      <li
        key="5"
      >
        Oranges
        <br />
        0.3
         
        1.99
        /
        kg
         
        0.6
      </li>
    </ul>
    <Totals>
      <div
        className="Totals"
      >
        <span>
          -----
        </span>
        <span
          className="label"
        >
          Sub-total
        </span>
        <span
          className="value sub-total"
        >
          3.5
        </span>
        <br />
        <span>
          Savings
        </span>
        <span
          className="savings"
          key="Beans 3 for 2"
        >
          <span>
            Beans 3 for 2
          </span>
          <span>
            -0.5
          </span>
        </span>
        <span
          className="savings"
          key="Coke 2 for £1"
        >
          <span>
            Coke 2 for £1
          </span>
          <span>
            -0.4
          </span>
        </span>
        <span>
          -----
        </span>
        <span
          className="label"
        >
          Total savings
        </span>
        <span
          className="value total-savings"
        >
          -0.9
        </span>
        <span>
          -------------------------
        </span>
        <span
          className="label"
        >
          Total to Pay
        </span>
        <span
          className="value total-to-pay"
        >
          2.60
        </span>
      </div>
    </Totals>
  </Receipt>
</Provider>
`;
